service/history/mutableStateBuilder.go

```diff
@@ -1280,6 +1359,65 @@ func (e *mutableStateBuilder) AddRequestCancelExternalWorkflowExecutionFailedEve
 	return nil
 }
 
+func (e *mutableStateBuilder) AddSignalExternalWorkflowExecutionInitiatedEvent(decisionCompletedEventID int64,
+	signalRequestID string, request *workflow.SignalExternalWorkflowExecutionDecisionAttributes) *workflow.HistoryEvent {
+
+	event := e.hBuilder.AddSignalExternalWorkflowExecutionInitiatedEvent(decisionCompletedEventID, request)
+	if event == nil {
+		return nil
+	}
+
+	initiatedEventID := *event.EventId
+	ri := &persistence.SignalInfo{
+		InitiatedID:     initiatedEventID,
+		SignalRequestID: signalRequestID,
+		SignalName:      request.GetSignalName(),
+		Input:           request.Input,
+	}
+
+	e.pendingSignalInfoIDs[initiatedEventID] = ri
+	e.updateSignalInfos = append(e.updateSignalInfos, ri)
+
+	return event
+}
+
+func (e *mutableStateBuilder) AddExternalWorkflowExecutionSignalRequested(initiatedID int64,
+	domain, workflowID, runID string) *workflow.HistoryEvent {
+	_, ok := e.GetSignalInfo(initiatedID)
+	if !ok {
+		logging.LogInvalidHistoryActionEvent(e.logger, logging.TagValueActionWorkflowSignalRequested, e.GetNextEventID(),
+			fmt.Sprintf("{InitiatedID: %v, Exist: %v}", initiatedID, ok))
+
+		return nil
+	}
+
+	if e.DeletePendingSignal(initiatedID) == nil {
```

Comment:
What if this is not nil?

Reply:
Good question. We have some options: 
1. add new `PendingSignalNotExistEvent` event when err happens
2. add success (or failed) event no matter err happens or not
3. add nothing when err happens (current implementation)
Which one do you like?

Reply:
I added log if err happens.